etl/steps/data/garden/eth/2023-03-15/ethnic_power_relations.py

```diff
@@ -0,0 +1,258 @@
+    df_regions["lmaxexclpop"] = df_regions["maxexclpop"] / df_regions["lpop"]
+
+    # Concatenate df with df_regions
+    df = pd.concat([df, df_regions], ignore_index=True)
+
+    # Drop headcount and population columns
+    df = df.drop(columns=["population"] + pop_vars_headcount)
+
+    # Verify index and sort
+    df = df.set_index(["country", "year"], verify_integrity=True).sort_index()
+
+    return df
+
+
+# EPR classifies some countries as both their historical and current name. This is not ideal for some cases where their borders changed completely
+def separate_historical_countries(df: pd.DataFrame) -> pd.DataFrame:
+
+    df["country"] = df["country"].astype(str)
+    # RUSSIA / SOVIET UNION
+    # Change the country name for the Soviet Union to Russia for years less or equal to 1991
+    df.loc[(df["country"] == "Russia (Soviet Union)") & (df["year"] <= 1991), "country"] = "USSR"
+    df.loc[(df["country"] == "Russia (Soviet Union)") & (df["year"] > 1991), "country"] = "Russia"
+
+    # VIETNAM
+    # Drop values for "Vietnam, Democratic Republic of" between years 1954 and 1975
+    df = df[
+        ~((df["country"] == "Vietnam, Democratic Republic of") & (df["year"] >= 1954) & (df["year"] <= 1975))
+    ].reset_index(drop=True)
+
+    # YEMEN
+    # Rename "Yemen (Arab Republic of Yemen)" to "Yemen" for years greater than 1990
+    df.loc[(df["country"] == "Yemen (Arab Republic of Yemen)") & (df["year"] > 1990), "country"] = "Yemen"
+
+    # GERMANY
+    # Rename "German Federal Republic" to "Germany" for years greater than 1990
+    df.loc[(df["country"] == "German Federal Republic") & (df["year"] > 1990), "country"] = "Germany"
+
+    return df
+
+
+def run(dest_dir: str) -> None:
+    log.info("ethnic_power_relations.start")
+
+    #
+    # Load inputs.
+    #
+    # Load meadow dataset.
+    ds_meadow: Dataset = paths.load_dependency("growup")
+
+    # Read table from meadow dataset.
+    tb_meadow = ds_meadow["growup"]
+
+    # Create a dataframe with data from the table.
+    df = pd.DataFrame(tb_meadow).reset_index()
+
+    #
+    # Process data.
+    #
+
+    # Separate historical countries (Soviet Union, Vietnam, Yemen)
+    df = separate_historical_countries(df)
+
+    log.info("ethnic_power_relations.harmonize_countries")
+    df = geo.harmonize_countries(df=df, countries_file=paths.country_mapping_path)
+
+    # Select and transform variables
+    df = process_ethnic_power_relations(df)
+
+    # Add regional aggregations
+    df = add_regional_aggregations(df)
+
+    # Create a new table with the processed data and add a new short_name.
+    tb_garden = Table(df)
+    tb_garden.metadata.short_name = "ethnic_power_relations"
+
+    #
+    # Save outputs.
+    #
+    # Create a new garden dataset with the same metadata as the meadow dataset.
+    ds_garden = create_dataset(dest_dir, tables=[tb_garden])
+
+    # For now the variable descriptions are stored as a list of strings, this transforms them into a single string
```

Comment:
I am using YAML anchors in the garden metadata

Reply:
Oh yeah, that's tricky. The reason is doesn't work is that even though calling `create_dataset` updates metadata of dataset and table, it does it on a copy of the table, so the input `tb_garden` won't be modified. You have to load it from dataset like this

```python
# Create a new garden dataset with the same metadata as the meadow dataset.
    ds_garden = create_dataset(dest_dir, tables=[tb_garden])

    # For now the variable descriptions are stored as a list of strings, this transforms them into a single string
    tb_garden = ds_garden["ethnic_power_relations"]
    for col in tb_garden.columns:
        if isinstance(tb_garden[col].metadata.description, list):
            tb_garden[col].metadata.description = "\n\n".join(tb_garden[col].metadata.description)
    ds_garden.add(tb_garden)
```

Even better would be to implement [placeholders in YAML files](https://github.com/owid/etl/issues/791) to avoid this hack with lists, but that ball is on my side.

Reply:
Oh, thank you! I actually tried something very similar (just replacing `tb_garden` to  `ds_garden["ethnic_power_relations"]` in the loop) and it failed, because it said that some other variable unit was empty. (I think I forgot to also change the `ds_garden.add(tb_garden)` to `ds_garden.add(ds_garden["ethnic_power_relations"])`!)